activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -445,8 +445,18 @@ def check_constraints(table_name)
             options = {
               name: row["name"]
             }
+
             expression = row["expression"]
-            expression = expression[1..-2] unless mariadb? # remove parentheses added by mysql
+
+            unless mariadb?
+              # remove parentheses added by MySQL
+              expression = expression[1..-2]
+
+              # MySQL returns check constraints expression in an already escaped form.
+              # This leads to duplicate escaping later (e.g. when the expression is used in the SchemaDumper).
+              expression = expression.gsub("\\'", "'")
```

Comment:
I'm a bit curious how these got here in the first place, but not sure that question is relevant to this bug. 🤔

Reply:
I actually don't know why MySQL does this. In fact, the `\'` finds its way into the MySQL information schema even without the use of Rails at all.

Running something like:

```sql
CREATE TABLE `posts` (`id` bigint NOT NULL AUTO_INCREMENT PRIMARY KEY, `title` varchar(255), CONSTRAINT chk_rails_fa84ad2808 CHECK ((title != 'forbidden_title')))
```

And then reading back the the created constraint with:

```sql
SELECT cc.constraint_name AS 'name',
       cc.check_clause AS 'expression'
FROM information_schema.check_constraints cc
         JOIN information_schema.table_constraints tc
              USING (constraint_schema, constraint_name)
WHERE tc.table_name = 'posts'
```

(which is what ActiveRecord does in `AbstractMysqlAdapter#check_constraints`)

shows the already escaped result on the database level:

![Bildschirmfoto 2021-06-10 um 17 01 20](https://user-images.githubusercontent.com/1724355/121548839-a4035700-ca0d-11eb-98fb-83b05f2dfc58.png)

There is a difference in string/quote handling in the MySQL world in contrast to the (ASCII) SQL standard world, as MySQL accepts backslashes as escape character. There even is a configuration value `NO_BACKSLASH_ESCAPES` for the sql_mode option of MySQL.

I tried setting it with

```sql
SET GLOBAL sql_mode = 'ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION,NO_BACKSLASH_ESCAPES';
```

But it leads to a totally different and weird error, when I try to reproduce the issue with the executable version of the problem. I was not yet able track it down (which would be unrelated to the bug here).

```
~/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.1.3.2/lib/active_record/connection_adapters/mysql2_adapter.rb:103:in `escape': negative string size (or size too big) (ArgumentError)
```

What I think is a bit weird is, that even when adding `NO_BACKSLASH_ESCAPES` MySQL still adds the `\` for escaping, but to be honest, I have no clue if that is a MySQL bug or desired behavior.

_Fun fact, I learned that `NO_BACKSLASH_ESCAPES` was added as as a special request from SAP https://bugs.mysql.com/bug.php?id=6368, https://github.com/mysql/mysql-server/blob/7ed30a748964c009d4909cb8b4b22036ebdef239/mysql-test/t/sql_mode.test#L85_